onadata/libs/serializers/merged_xform_serializer.py

```diff
@@ -179,4 +178,13 @@ def create(self, validated_data):
             __.instances_with_geopoints for __ in validated_data.get('xforms')
         ])
 
-        return super(MergedXFormSerializer, self).create(validated_data)
+        with transaction.atomic():
+            instance = super(MergedXFormSerializer,
+                             self).create(validated_data)
+
+        if instance.xforms.all().count() != len(xforms):
```

Comment:
is this to guard against some synchronization problem? Like if a new xform is created by another user on line 184.

Either way, it seems like this is something we, as the server might want to retry a couple times (with increasing pause times) before returning a validation error to the client

Reply:
The serializer already validates we have the correct number of forms to create a merged dataset, but due to the nature of how Django handles many to many relationships, we have had incidences in preview where a merged dataset was created but the forms were not linked. I have yet to find a way to replicate the issue, a check after the fact I hope will catch the issue. Nothing to do with user synchronization.

Reply:
Cool, then I think we should retry this in the code a couple times with back-off before returning an error to the user

Reply:
I think I will take the approach of adding the forms again, it's not clear what/why it failed in the `create()` method. Retrying the whole function might create more objects that are not necessary. The idea of the exception was to force a rollback on the request but I suppose we will have to wait and see if it will reoccur and hopefully have a better understanding of why it is behaving this way.